File: application/types/lesson.ts
export interface Lesson {
  id: number;
  subject_id: number | null;
  subject_name: string | null;
  title: string;
  description?: string;
  day_of_week: number; // 0 = Sunday, 1 = Monday, ..., 6 = Saturday
  day_name: string;
  start_time: string; // "HH:mm" e.g. "10:00"
  end_time: string | null; // "HH:mm" e.g. "11:30"
  location: string;
  color: string;
  reminder_minutes: number;
  created_at: string;
  updated_at: string;
}

export interface LessonInput {
  subject_id?: number | null;
  title: string;
  description?: string;
  day_of_week?: number;
  days_of_week?: number[];
  start_time: string; // "HH:mm" e.g. "10:00"
  end_time?: string | null;
  location?: string;
  color?: string;
  reminder_minutes?: number;
}

export interface DayOfWeekOption {
  value: number;
  label: string;
  short: string;
}

export const DAYS_OF_WEEK: DayOfWeekOption[] = [
  { value: 1, label: "Monday", short: "Mon" },
  { value: 2, label: "Tuesday", short: "Tue" },
  { value: 3, label: "Wednesday", short: "Wed" },
  { value: 4, label: "Thursday", short: "Thu" },
  { value: 5, label: "Friday", short: "Fri" },
  { value: 6, label: "Saturday", short: "Sat" },
  { value: 0, label: "Sunday", short: "Sun" },
];

export interface ReminderOption {
  value: number;
  label: string;
}

export const REMINDER_OPTIONS: ReminderOption[] = [
  { value: 0, label: "At start of lesson" },
  { value: 5, label: "5 minutes before" },
  { value: 10, label: "10 minutes before" },
  { value: 15, label: "15 minutes before" },
  { value: 30, label: "30 minutes before" },
  { value: 45, label: "45 minutes before" },
  { value: 60, label: "1 hour before" },
  { value: 120, label: "2 hours before" },
  { value: 180, label: "3 hours before" },
  { value: 240, label: "4 hours before" },
  { value: 360, label: "6 hours before" },
  { value: 480, label: "8 hours before" },
  { value: 720, label: "12 hours before" },
  { value: 1440, label: "1 day before" },
  { value: 2880, label: "2 days before" },
  { value: 4320, label: "3 days before" },
];

File: backend/internal/dto/lesson.go
package dto

import (
	"time"

	"github.com/abde1khaliq/korasa/internal/models"
)

var dayNames = [7]string{
	"Sunday",
	"Monday",
	"Tuesday",
	"Wednesday",
	"Thursday",
	"Friday",
	"Saturday",
}

type LessonResponse struct {
	ID              int     `json:"id"`
	SubjectID       *int    `json:"subject_id"`
	SubjectName     *string `json:"subject_name"`
	Title           string  `json:"title"`
	Description     string  `json:"description"`
	DayOfWeek       int     `json:"day_of_week"`
	DayName         string  `json:"day_name"`
	StartTime       string  `json:"start_time"`
	EndTime         *string `json:"end_time"`
	Location        string  `json:"location"`
	Color           string  `json:"color"`
	ReminderMinutes int     `json:"reminder_minutes"`
	CreatedAt       string  `json:"created_at"`
	UpdatedAt       string  `json:"updated_at"`
}

// ToLessonResponse transforms a Lesson model into a client-safe DTO.
func ToLessonResponse(l models.Lesson) LessonResponse {
	var subjectName *string
	if l.Subject != nil {
		subjectName = &l.Subject.Name
	}

	dayName := "Unknown"
	if l.DayOfWeek >= 0 && l.DayOfWeek < 7 {
		dayName = dayNames[l.DayOfWeek]
	}

	return LessonResponse{
		ID:              l.ID,
		SubjectID:       l.SubjectID,
		SubjectName:     subjectName,
		Title:           l.Title,
		Description:     l.Description,
		DayOfWeek:       l.DayOfWeek,
		DayName:         dayName,
		StartTime:       l.StartTime,
		EndTime:         l.EndTime,
		Location:        l.Location,
		Color:           l.Color,
		ReminderMinutes: l.ReminderMinutes,
		CreatedAt:       l.CreatedAt.Format(time.RFC3339),
		UpdatedAt:       l.UpdatedAt.Format(time.RFC3339),
	}
}

// ToLessonListResponse transforms a slice of Lesson models into a slice of DTOs.
func ToLessonListResponse(lessons []models.Lesson) []LessonResponse {
	resp := make([]LessonResponse, len(lessons))
	for i, l := range lessons {
		resp[i] = ToLessonResponse(l)
	}
	return resp
}
